### PR TITLE
Ensure principal axes follow right hand convention 

### LINKED ISSUE #2637: Principal axes have inconsistent handedness and signs
I was comparing the behaviour of the orient package in vmd and the align_to_principal_axis() in mdanalysis and i noticed that I was getting different poses. 

When I print out the principal_axes() of my universe I notice that the values are correct, and the basis is orthogonal but they have inconsistent signs, looking at the code I'm not sure why otherwise I would fix it. 

proof.
pdb 1unp
program mdanalysis
[[-0.16834601 -0.97931099  0.11229246]
[ 0.96713188 -0.14207155  0.21088289]
[-0.19056637  0.14410291  0.97104006]]

program vmd 
0.16834793174985047 0.9793106237038721 -0.11229281444634183
-0.967131578444061 0.1420733560320056 -0.21088307538089016
-0.19056621907317922 0.14410365646483728 0.971039984934512

pdb 6msm
program mdanalysis
[[ 0.89960788 -0.43591176 -0.02620299]
 [-0.43650301 -0.89938041 -0.02408336]
  [ 0.01306824 -0.03310326  0.9993665 ]]

program vmd
0.8996078490065614 -0.43591182912687276 -0.026202962295357412
0.4365030807649993 0.8993803722626518 0.02408332351147616
0.013068224380240984 -0.03310322062818084 0.9993664984857106

### PR DESCRIPTION
**Pull request by WG150**
Fixes #2637 

Changes made in this Pull Request:
 - Added a check when calculating the principal axes to ensure that they follow the right handed convention.
- Added a test with PDB code 6msm that returned a triple product of -1.0. Now returns 1.0

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/CHANGELOG [hunk 1]
```diff
@@ -22,6 +22,7 @@ mm/dd/yy richardjgowers, kain88-de, lilyminium, p-j-smith, bdice, joaomcteixeira
   * 0.21.0
 
 Fixes
+  * Ensures principa_axes() returns axes with the right hand convention (Issue #2637)
```
Comment (RMeli) [inline package/CHANGELOG:25]: Typo on function name?

File: testsuite/MDAnalysisTests/core/test_topologyattrs.py [hunk 2]
```diff
@@ -335,6 +335,11 @@ def test_principal_axes(self, ag):
                       [1.20986911e-02, 9.98951474e-01, -4.41539838e-02],
                       [-9.99925632e-01, 1.21546132e-02, 9.98264877e-04]]))
 
+    def test_principal_axes_handedness(self):
+        universe = mda.Universe(PDB_CHECK_RIGHTHAND_PA)
```
Comment (RMeli) [inline testsuite/MDAnalysisTests/core/test_topologyattrs.py:339]: To anticipate @orbeckst comments, I think this is better off as a [fixture](https://docs.pytest.org/en/latest/fixture.html#fixtures-as-function-arguments). Something like
```
@pytest.fixture
def universe_pa():
    return mda.Universe(PDB_CHECK_RIGHTHAND_PA)

def test_principal_axes_handedness(self, universe_pa):
    pass
```

File: package/MDAnalysis/core/topologyattrs.py [hunk 3]
```diff
@@ -1039,11 +1039,17 @@ def principal_axes(group, pbc=False):
...
+        
         # Sort
         indices = np.argsort(e_val)[::-1]
-        # Return transposed in more logical form. See Issue 33.
-        return e_vec[:, indices].T
+        # Make transposed in more logical form. See Issue 33.
+        e_vec = e_vec[:, indices].T
+
+        # Make sure the right hand convention is followed
+        if np.dot(np.cross(e_vec[0], e_vec[1]), e_vec[2]) == -1.0:
```
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py:1049]: Don’t check floats for exact equality. Better just check if < 0

File: package/MDAnalysis/core/topologyattrs.py [hunk 4]
```diff
@@ -1039,11 +1039,17 @@ def principal_axes(group, pbc=False):
...
+            e_vec = e_vec * -1
```
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py:1050]: Use infix notation for more conciseness: `e_vec *= -1`

File: testsuite/MDAnalysisTests/datafiles.py [hunk 5]
```diff
@@ -186,7 +186,8 @@
     "GMX_DIR", # GROMACS directory
     "GMX_TOP_BAD", # file with an #include that doesn't exist
     "ITP_no_endif", # file missing an #endif
-    "PDB_CRYOEM_BOX" #Issue 2599
+    "PDB_CRYOEM_BOX", #Issue 2599
+    "PDB_CHECK_RIGHTHAND_PA" # for testing right handedness of principal_axes
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/datafiles.py:190]: Store the file as bzip2 compressed.

File: package/MDAnalysis/core/topologyattrs.py (hunk 3 above)
Comment (WG150) [reply package/MDAnalysis/core/topologyattrs.py:1049]: Sure - that makes sense, this was actually already pointed out to me by @RMeli in a private chat.

File: testsuite/MDAnalysisTests/core/test_atomgroup.py [hunk 6]
```diff
@@ -941,9 +941,9 @@ def ref_noPBC(self):
                                 dtype=np.float32),
             'BSph': (173.40482, np.array([4.23789883, 0.62429816, 2.43123484], dtype=np.float32)),
             'PAxes': np.array([
-                [-0.78787867, -0.26771575, 0.55459488],
-                [0.40611024, 0.45112859, 0.7947059],
-                [0.46294889, -0.85135849, 0.24671249]])
+                [ 0.78787867,  0.26771575, -0.55459488],
```
Comment (WG150) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py:944]: Updated the test PAxes to an array that returns 1 when checking the triple product via:
 `np.dot(np.cross(PAxes[0], PAxes[1]), PAxes[2])`

The original returned -1 (as I assume handedness wasn't checked before?).

File: testsuite/MDAnalysisTests/core/test_atomgroup.py [hunk 7]
```diff
@@ -962,9 +962,9 @@ def ref_PBC(self):
                 dtype=np.float32),
             'BSph': (47.923367, np.array([26.82960892, 31.5592289, 30.98238945], dtype=np.float32)),
             'PAxes': np.array([
-                [-0.85911708, 0.19258726, 0.4741603],
-                [-0.07520116, -0.96394227, 0.25526473],
-                [-0.50622389, -0.18364489, -0.84262206]])
+                [ 0.85911708, -0.19258726, -0.4741603 ],
```
Comment (WG150) [inline testsuite/MDAnalysisTests/core/test_atomgroup.py:965]: Again, 

Updated the test PAxes to an array that returns 1. The original returned -1.

File: testsuite/MDAnalysisTests/datafiles.py (hunk 5 above)
Comment (WG150) [reply testsuite/MDAnalysisTests/datafiles.py:190]: This has now been done.

File: package/MDAnalysis/core/topologyattrs.py [hunk 8]
```diff
@@ -1039,11 +1039,17 @@ def principal_axes(group, pbc=False):
...
         # Sort
         indices = np.argsort(e_val)[::-1]
-        # Return transposed in more logical form. See Issue 33.
-        return e_vec[:, indices].T
+        # Make transposed in more logical form. See Issue 33.
+        e_vec = e_vec[:, indices].T
+
+        # Make sure the right hand convention is followed
+        if np.dot(np.cross(e_vec[0], e_vec[1]), e_vec[2]) < 0:
+            e_vec *= -1
```
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py:1050]: Multiply all of them or only `evec[2]`  – either one works, not sure if one is preferable over the other.

File: package/MDAnalysis/core/topologyattrs.py [hunk 9]
```diff
@@ -1039,11 +1039,17 @@ def principal_axes(group, pbc=False):
         """
         atomgroup = group.atoms
         e_val, e_vec = np.linalg.eig(atomgroup.moment_of_inertia(pbc=pbc))
-
+
```
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py:1042]: Forgot in my previous review: please add a `.. versionchanged: 1.0.0` entry that states that this function is now guaranteed to return evecs in a R-handed coordinate system. Add a similar sentence to the docs themselves. This is a change that users need to be aware of.

Comment (RMeli) [reply package/MDAnalysis/core/topologyattrs.py:1042]: Very pedantic, but while you are at it: PEP8 requires blank lines to be empty (line 1042).

File: package/CHANGELOG [hunk 10]
```diff
@@ -22,6 +22,7 @@ mm/dd/yy richardjgowers, kain88-de, lilyminium, p-j-smith, bdice, joaomcteixeira
   * 0.21.0
 
 Fixes
+  * Ensures principal_axes() returns axes with the right hand convention (Issue #2637)
```
Comment (RMeli) [inline package/CHANGELOG:25]: You should add your name/GitHub handle in the `CHANGELOG` as well.

File: package/MDAnalysis/core/topologyattrs.py (hunk 9 above)
Comment (WG150) [reply package/MDAnalysis/core/topologyattrs.py]: Added to the doc and sorted line 1042

File: package/CHANGELOG (hunk 10 above)
Comment (WG150) [reply package/CHANGELOG:25]: Done! Thanks for reminding me.

File: package/MDAnalysis/core/topologyattrs.py [hunk 11]
```diff
@@ -1035,11 +1035,13 @@ def principal_axes(group, pbc=False):
 
 
         .. versionchanged:: 0.8 Added *pbc* keyword
+        .. versionchanged:: 1.0.0 
+        Always return principal axes in right hand convention.
```
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py:1039]: Needs to be indented properly
```suggestion
           Always return principal axes in right hand convention.
```

File: package/MDAnalysis/core/topologyattrs.py [hunk 12]
```diff
@@ -1035,11 +1035,13 @@ def principal_axes(group, pbc=False):
 
 
         .. versionchanged:: 0.8 Added *pbc* keyword
+        .. versionchanged:: 1.0.0
```
Comment (orbeckst) [inline package/MDAnalysis/core/topologyattrs.py:1038]: Also add text to the doc string of the method saying that the eigenvectors form a right-handed coordinate system.

Comment (WG150) [reply package/MDAnalysis/core/topologyattrs.py:1038]: Sorry, must've missed this. Now added.

**Comment by orbeckst** [comment]
@RMeli thanks for the review!

**Comment by orbeckst** [comment]
@WG150 looking good, now just waiting for CI to come back all ok and then I'll squash-merge.

**Comment by orbeckst** [comment]
Well done, thank you very much for your contribution @WG150 .

I'll send you the GH invite as a contributor.

**Comment by richardjgowers** [comment]
@WG150 welcome aboard!

**Pull request merged**